File: music_sync.rb
#!/usr/bin/env ruby

require 'rubygems'
require 'json'
require 'net/http'
require 'uri'
require 'time'
require 'fileutils'

require 'appscript'
include Appscript

class MusicSync
  def initialize
    @config = load_config
    setup_menu_bar
  end

  def run
    # Keep the app running
    loop do
      sleep 1
    end
  end

  private

  def load_config
    config_path = File.expand_path('~/.music-sync-config.json')
    
    unless File.exist?(config_path)
      default_config = {
        'source_path' => File.expand_path('~/Music/Swinsian'),
        'server_address' => 'Daves-Mac-mini.local',
        'server_user' => 'dave',
        'server_path' => '/Volumes/DATA/Music Library',
        'plex_token' => '', # Optional: Your Plex token for API access
        'plex_server' => 'http://Daves-mac-mini.local:32400' # Your Plex server address
      }
      
      File.write(config_path, JSON.pretty_generate(default_config))
      return default_config
    end
    
    JSON.parse(File.read(config_path))
  end

  def setup_menu_bar
    app = app('System Events').processes['SystemUIServer']
    
    # Create menu bar extra
    @menu_bar = app.menu_bars[1].menu_bar_items.end.make(
      :new => :menu_bar_item,
      :with_properties => { :title => '🎵' }
    )
    
    # Create menu
    menu = @menu_bar.menus[1].make(:new => :menu)
    
    # Add menu items
    @sync_item = menu.menu_items.end.make(
      :new => :menu_item,
      :with_properties => { :title => 'Sync Now' }
    )
    
    @last_sync_item = menu.menu_items.end.make(
      :new => :menu_item,
      :with_properties => { :title => 'Last Sync: Never', :enabled => false }
    )
    
    menu.menu_items.end.make(
      :new => :menu_item,
      :with_properties => { :title => '-' } # Separator
    )
    
    @settings_item = menu.menu_items.end.make(
      :new => :menu_item,
      :with_properties => { :title => 'Settings...' }
    )
    
    # Set up click handlers
    @sync_item.clicked.connect { sync_library }
    @settings_item.clicked.connect { show_settings }
  end

  def sync_library
    begin
      @sync_item.title.set('Syncing...')
      @sync_item.enabled.set(false)
      
      # Run rsync
      rsync_cmd = [
        'rsync',
        '-av',
        '--delete',
        "#{@config['source_path']}/",
        "#{@config['server_user']}@#{@config['server_address']}:#{@config['server_path']}"
      ]
      
      result = system(*rsync_cmd)
      raise "rsync failed" unless result
      
      # Trigger Plex scan if configured
      trigger_plex_scan if @config['plex_token'].to_s.length > 0
      
      # Update last sync time
      now = Time.now.strftime('%Y-%m-%d %H:%M')
      @last_sync_item.title.set("Last Sync: #{now}")
      
      # Show notification
      app('System Events').processes['NotificationCenter'].notifications.end.make(
        :new => :notification,
        :with_properties => {
          :title => 'Music Sync Complete',
          :subtitle => 'Library successfully synchronized',
          :message => "Synced at #{now}"
        }
      )
      
    rescue => e
      app('System Events').processes['NotificationCenter'].notifications.end.make(
        :new => :notification,
        :with_properties => {
          :title => 'Music Sync Failed',
          :subtitle => 'Error during synchronization',
          :message => e.message
        }
      )
      
    ensure
      @sync_item.title.set('Sync Now')
      @sync_item.enabled.set(true)
    end
  end

  def trigger_plex_scan
    return unless @config['plex_token']
    
    # Get music library section ID
    uri = URI("#{@config['plex_server']}/library/sections")
    req = Net::HTTP::Get.new(uri)
    req['X-Plex-Token'] = @config['plex_token']
    
    res = Net::HTTP.start(uri.hostname, uri.port) do |http|
      http.request(req)
    end
    
    raise "Failed to get Plex sections" unless res.is_a?(Net::HTTPSuccess)
    
    # Parse JSON response and find music library
    sections = JSON.parse(res.body)
    music_section = sections['MediaContainer']['Directory'].find { |dir| dir['type'] == 'artist' }
    raise "No music library found in Plex" unless music_section
    
    # Trigger scan
    uri = URI("#{@config['plex_server']}/library/sections/#{music_section['key']}/refresh")
    req = Net::HTTP::Get.new(uri)
    req['X-Plex-Token'] = @config['plex_token']
    
    res = Net::HTTP.start(uri.hostname, uri.port) do |http|
      http.request(req)
    end
    
    raise "Failed to trigger Plex scan" unless res.is_a?(Net::HTTPSuccess)
  end

  def show_settings
    config_path = File.expand_path('~/.music-sync-config.json')
    app('System Events').processes['NotificationCenter'].notifications.end.make(
      :new => :notification,
      :with_properties => {
        :title => 'Music Sync Settings',
        :subtitle => 'Edit configuration file:',
        :message => config_path
      }
    )
  end
end

# Start the app
MusicSync.new.run

File: music_sync.py
#!/usr/bin/env python3
import rumps
import subprocess
import os
import json
from datetime import datetime

class MusicSyncApp(rumps.App):
    def __init__(self):
        super(MusicSyncApp, self).__init__("🎵")
        
        # Load config
        self.config = self.load_config()
        
        # Menu items
        self.sync_button = rumps.MenuItem("Sync Now")
        self.last_sync = rumps.MenuItem("Last Sync: Never")
        self.settings = rumps.MenuItem("Settings...")
        
        # Add menu items
        self.menu = [
            self.sync_button,
            self.last_sync,
            None,  # Separator
            self.settings
        ]

    @rumps.clicked("Sync Now")
    def sync(self, _):
        try:
            # Start sync
            self.sync_button.title = "Syncing..."
            self.sync_button.set_callback(None)  # Disable during sync
            
            # Run rsync command
            rsync_cmd = [
                "rsync",
                "-av",
                "--delete",
                self.config["source_path"] + "/",
                f"{self.config['server_user']}@{self.config['server_address']}:{self.config['server_path']}"
            ]
            
            result = subprocess.run(rsync_cmd, capture_output=True, text=True)
            
            if result.returncode != 0:
                raise Exception(f"rsync failed: {result.stderr}")
            
            # Trigger Plex scan via API
            if self.config.get("plex_token"):
                self.trigger_plex_scan()
            
            # Update last sync time
            now = datetime.now().strftime("%Y-%m-%d %H:%M")
            self.last_sync.title = f"Last Sync: {now}"
            
            # Show success notification
            rumps.notification(
                title="Music Sync Complete",
                subtitle="Library successfully synchronized",
                message=f"Synced at {now}"
            )
            
        except Exception as e:
            rumps.notification(
                title="Music Sync Failed",
                subtitle="Error during synchronization",
                message=str(e)
            )
        
        finally:
            # Re-enable sync button
            self.sync_button.title = "Sync Now"
            self.sync_button.set_callback(self.sync)

    @rumps.clicked("Settings...")
    def settings_dialog(self, _):
        # Note: In a full implementation, this would open a proper settings window
        # For now, we'll just show where to edit the config file
        config_path = os.path.expanduser("~/.music-sync-config.json")
        rumps.notification(
            title="Music Sync Settings",
            subtitle="Edit configuration file:",
            message=config_path
        )

    def load_config(self):
        """Load configuration from JSON file"""
        config_path = os.path.expanduser("~/.music-sync-config.json")
        
        if not os.path.exists(config_path):
            # Create default config
            default_config = {
                "source_path": os.path.expanduser("~/Music"),
                "server_address": "mini.local",
                "server_user": "your_username",
                "server_path": "/volume1/music",
                "plex_token": "",  # Optional: Your Plex token for API access
                "plex_server": "http://mini.local:32400"  # Your Plex server address
            }
            
            with open(config_path, 'w') as f:
                json.dump(default_config, f, indent=4)
            
            return default_config
        
        with open(config_path, 'r') as f:
            return json.load(f)

    def trigger_plex_scan(self):
        """Trigger a Plex library scan via API"""
        if not self.config.get("plex_token"):
            return
            
        import requests
        
        headers = {
            'X-Plex-Token': self.config['plex_token']
        }
        
        # Get music library section ID
        sections_url = f"{self.config['plex_server']}/library/sections"
        response = requests.get(sections_url, headers=headers)
        response.raise_for_status()
        
        # Find music library section
        for section in response.json()['MediaContainer']['Directory']:
            if section['type'] == 'artist':
                section_id = section['key']
                break
        else:
            raise Exception("No music library found in Plex")
        
        # Trigger scan
        scan_url = f"{self.config['plex_server']}/library/sections/{section_id}/refresh"
        response = requests.get(scan_url, headers=headers)
        response.raise_for_status()

if __name__ == "__main__":
    MusicSyncApp().run()